### PR TITLE
Grid Overhaul 

### PR DESCRIPTION
**Pull request by VHchavez**
## Description
All changes are located in `n2v/grid/grider.py`
Quantities on the grid can now be obtained individually. [Density, Orbitals, Hartree/ESP/External, Vxc]
Each function requires either a grid or V_potential so that each function can either work for 1d, cubic, and DFT grid. 
Adds docstring for each function. 

This is a massive PR( :0 ), but all functions follow the same structure. So by focusing on one, we should decide if it's a good way to go.  The main idea is that the same function can be used to extract the required grid. e.g.:

```
npoints = 100
x = np.linspace(-2,2,npoints)[:,None]
y = np.linspace(-2,2,npoints)[:,None]
z = np.linspace(-2,2,npoints)[:,None] 
grid = np.concatenate((x,y,z), axis=1).T
density = inverter.on_grid_density(grid=grid)

print("Normal Grid")
plt.plot(density[:,0])
plt.show()

print("Cubic Grid")
density = inverter.on_grid_density(grid=grid, cubic_grid=True)
plt.imshow(density[:,:,5,0], interpolation="bicubic")
plt.show()


print("DFT GRID")
grid = inverter.generate_dft_grid(wfn)
density = inverter.on_grid_density(vpot=wfn.V_potential())
plt.scatter(grid[0], density[:,0])
plt.show()
```
To produce:

image

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [ x] Thoroughly test functions. Will address in future PR testing all classes. 

## Questions
- [ ] ??

## Status
- [x] Ready to go

### CODE REVIEW

**Comment by VHchavez** [comment]
What do you think? @ymshi449

File: n2v/grid/grider.py [hunk 1]
```diff
@@ -0,0 +1,918 @@
...
+        offset = 0
+        for i_block in blocks:
+            points_function.compute_points(i_block)
+            b_points = i_block.npoints()
+            offset += b_points
+            w = np.array(i_block.w())
+            lpos = np.array(i_block.functions_local_to_global())
+            phi = np.array(points_function.basis_values()["PHI"])[:b_points, :lpos.shape[0]]
+
+            l_mat = matrix[(lpos[:, None], lpos)]
```
Comment (ymshi449) [inline n2v/grid/grider.py:215]: Here you already assumed matrix is of ndim==2. You should put this code in the following if else.

File: n2v/grid/grider.py [hunk 2]
```diff
@@ -0,0 +1,918 @@
...
+
+            if len(matrix.shape) == 1:
```
Comment (ymshi449) [inline n2v/grid/grider.py:217]: A better way here is is matrix.ndim == 2

File: n2v/grid/grider.py [hunk 3]
```diff
@@ -0,0 +1,918 @@
...
+    #Quantities on Grid
+    def on_grid_ao(self, matrix, grid=None, vpot=None, cubic_grid=False):
+        """
+        Generates a quantity on the grid given its ao representation
+
+        Parameters
+        ----------
+        matrix: np.ndarray
+            Quantity on ao basis
+        grid: np.ndarray
```
Comment (ymshi449) [inline n2v/grid/grider.py:171]: Always specify the shape of array. Here it should be (3, total_num_grid_points) or (4, total_num_grid_points)

File: n2v/grid/grider.py [hunk 4]
```diff
@@ -0,0 +1,918 @@
...
+            Grid where density will be computed
+        cubic_grid: bool    
+            Is the grid meant for a volume plot?
```
Comment (ymshi449) [inline n2v/grid/grider.py:174]: If yes, what would happen?

File: n2v/grid/grider.py [hunk 5]
```diff
@@ -0,0 +1,918 @@
...
+        vpot: psi4.core.VBase
+            Vpotential object with info about grid
```
Comment (ymshi449) [inline n2v/grid/grider.py:176]: Specify Vpotential will only play a role when grid is not given, in which case will return the result on the DFT spherical grid.

File: n2v/grid/grider.py [hunk 6]
```diff
@@ -0,0 +1,918 @@
...
+        elif cubic_grid is True and grid is None:
+            raise ValueError("'Cubic Grid' requires to explicitly specify grid")
+
+        if vpot is None and grid is not None:
+            blocks, npoints, points_function = self.grid_to_blocks(grid)
+        elif vpot is not None and grid is None:
+            nblocks = vpot.nblocks()
+            blocks = [vpot.get_block(i) for i in range(nblocks)]
+            npoints = vpot.grid().npoints()
+            points_function = vpot.properties()[0]
```
Comment (ymshi449) [inline n2v/grid/grider.py:197]: What if they are both not none? I think grid should have the priority than Vpot.

File: n2v/grid/grider.py [hunk 7]
```diff
@@ -0,0 +1,918 @@
...
+
+
+    #Quantities on Grid
+    def on_grid_ao(self, matrix, grid=None, vpot=None, cubic_grid=False):
+        """
+        Generates a quantity on the grid given its ao representation
+
+        Parameters
+        ----------
+        matrix: np.ndarray
```
[the remark this answers is not in the document]
Comment (ymshi449) [reply n2v/grid/grider.py:169]: I think coeff is better than matrix because it can be 1D or 2D. Matrix=2D.

File: n2v/grid/grider.py [hunk 8]
```diff
@@ -0,0 +1,918 @@
...
+
+        #Generate 2D information
+        if grid_type == "cubic":
+            self.c_grid.density = np.reshape(density, (N0, N1, N2, density.shape[1]))
+            self.c_grid.vext = np.reshape(density, (N0, N1, N2, vext.shape[1]))
+            self.c_grid.hartree = np.reshape(density, (N0, N1, N2, hartree.shape[1]))
+            self.c_grid.esp = np.reshape(density, (N0, N1, N2, esp.shape[1]))
+            self.c_grid.vxc = np.reshape(density, (N0, N1, N2, vxc.shape[1]))
+                
+    def build_rectangular_grid(self,
```
Comment (ymshi449) [inline n2v/grid/grider.py:663]: Do we still need this function? I thought its functional is totally replaced by above functions.

File: n2v/grid/grider.py [hunk 9]
```diff
@@ -0,0 +1,648 @@
...
+            raise ValueError("'Cubic Grid' requires to explicitly specify grid")
+
+        if grid is not None and vpot is None:
+            blocks, npoints, points_function = self.grid_to_blocks(grid)
+        elif grid is None and vpot is not None:
+            nblocks = vpot.nblocks()
+            blocks = [vpot.get_block(i) for i in range(nblocks)]
+            npoints = vpot.grid().npoints()
+            points_function = vpot.properties()[0]
+        elif grid is None and vpot is None:
```
Comment (ymshi449) [inline n2v/grid/grider.py:206]: Could you change this line just to "else:" to handle cases both None and both given?

**Review by ymshi449 — APPROVED** [review]

**Comment by VHchavez** [comment]
@ymshi449 I have addressed and fixed all of the improvements suggested. 
Let me know if you think anything else could be improved.  :)

**Review by ymshi449 — APPROVED** [review]

**Pull request merged**